src/array_api_extra/_lib/_funcs.py

```diff
@@ -28,6 +36,168 @@
+    function only for their matching condition, if the backend allows for it.
+
+    Examples
+    --------
+    >>> a = xp.asarray([5, 4, 3])
+    >>> b = xp.asarray([0, 2, 2])
+    >>> def f(a, b):
+    ...     return a // b
+    >>> apply_where(b != 0, f, a, b, fill_value=xp.nan)
+    array([ nan,  2., 1.])
+    """
+    # Parse and normalize arguments
+    mutually_exc_msg = "Exactly one of `fill_value` or `f2` must be given."
+    if is_array_api_obj(f2):
+        args = (cast(Array, f2), *args)
+        if fill_value is not None:
+            raise TypeError(mutually_exc_msg)
+        f2_: Callable[..., Array] | None = None  # type: ignore[no-any-explicit]
+    else:
+        if not callable(f2):
+            msg = "Third parameter must be either an Array or callable."
+            raise ValueError(msg)
+        f2_ = cast(Callable[..., Array], f2)  # type: ignore[no-any-explicit]
+        if fill_value is None:
+            raise TypeError(mutually_exc_msg)
+        if getattr(fill_value, "ndim", 0) != 0:
+            msg = "`fill_value` must be a scalar."
+            raise ValueError(msg)
+    del f2
+    if not args:
+        msg = "Must give at least one input array."
+        raise TypeError(msg)
+
+    xp = array_namespace(cond, *args) if xp is None else xp
+
+    # Determine output dtype
+    metas = [get_meta(arg, xp=xp) for arg in args]
+    temp1 = f1(*metas)
+    if f2_ is None:
+        if xp.__array_api_version__ >= "2024.12" or is_array_api_obj(fill_value):
+            dtype = xp.result_type(temp1.dtype, fill_value)
+        else:
+            # TODO: remove this when all backends support Array API 2024.12
+            dtype = (xp.empty((), dtype=temp1.dtype) * fill_value).dtype
+    else:
+        temp2 = f2_(*metas)
+        dtype = xp.result_type(temp1, temp2)
+
+    if is_dask_namespace(xp):
+        # Dask does not support assignment by boolean mask
+        meta_xp = array_namespace(get_meta(cond), *metas)
+        # pass dtype to both da.map_blocks and _apply_where
+        return xp.map_blocks(
+            partial(_apply_where, dtype=dtype, xp=meta_xp),
+            cond,
+            f1,
+            f2_,
+            *args,
+            fill_value=fill_value,
+            dtype=dtype,
+            meta=metas[0],
+        )
+
+    return _apply_where(cond, f1, f2_, *args, fill_value=fill_value, dtype=dtype, xp=xp)
+
+
+def _apply_where(  # type: ignore[no-any-explicit]  # numpydoc ignore=PR01,RT01
+    cond: Array,
+    f1: Callable[..., Array],
+    f2: Callable[..., Array] | None,
+    *args: Array,
+    fill_value: Array | int | float | complex | bool | None,
+    dtype: DType,
+    xp: ModuleType,
+) -> Array:
+    """Helper of `apply_where`. On Dask, this runs on a single chunk."""
+
+    if is_jax_namespace(xp):
+        # jax.jit does not support assignment by boolean mask
+        return xp.where(cond, f1(*args), f2(*args) if f2 is not None else fill_value)
```

Comment:
JAX-on-dask is currently unsupported by Dask. This is here and not much higher above only for this reason.
We'll need to add explicit unit tests when it lands in the future.